gateway/src/config.rs

```diff
@@ -0,0 +1,535 @@
+use std::{
+    fmt::{Display, Formatter, Result as FmtResult},
+    sync::Arc,
+};
+use twilight_gateway_queue::{LocalQueue, Queue};
+use twilight_model::gateway::{
+    payload::outgoing::{identify::IdentifyProperties, update_presence::UpdatePresencePayload},
+    Intents,
+};
+
+/// Identifier of a [shard], including the shard's ID and the total number of
+/// shards in use by the bot.
+///
+/// [shard]: crate::Shard
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct ShardId {
+    /// Current ID of the shard, 0-indexed.
+    current: u64,
+    /// Total number of shards used by the bot, 1-indexed.
+    total: u64,
+}
+
+impl ShardId {
+    /// ID of a bot that has only one shard.
+    ///
+    /// This should *only* be used by small bots in under one or two thousand
+    /// guilds.
+    pub const ONE: ShardId = ShardId::new(0, 1);
+
+    /// Create a new identifier for a shard.
+    ///
+    /// The current shard is 0-indexed while the total number of shards is
+    /// 1-indexed. This means that a current shard of 7 with a total of 8 is
+    /// valid, while a current shard value of 8 out of 8 total shards is
+    /// invalid.
+    ///
+    /// # Examples
+    ///
+    /// Create a new shard with a current index of 13 out of 24 shards:
+    ///
+    /// ```
+    /// use twilight_gateway::ShardId;
+    ///
+    /// let id = ShardId::new(13, 24);
+    /// ```
+    ///
+    /// # Panics
+    ///
+    /// Panics if the current shard is greater than or equal to the total number
+    /// of shards, or if the total number of shards is zero.
+    pub const fn new(current: u64, total: u64) -> Self {
+        assert!(total > 0, "total must be greater than zero");
+        assert!(
+            current < total,
+            "current shard (0-indexed) must be less than total (1-indexed)",
+        );
+
+        Self { current, total }
+    }
+
+    /// Create a new identifier for a shard if the shard indexes are valid.
+    ///
+    /// The current shard is 0-indexed while the total number of shards is
+    /// 1-indexed. This means that a current shard of 7 with a total of 8 is
+    /// valid, while a current shard value of 8 out of 8 total shards is
+    /// invalid.
+    pub const fn new_checked(current: u64, total: u64) -> Option<Self> {
+        let is_total_nonzero = total > 0;
+        let is_current_valid = current < total;
+
+        if is_total_nonzero && is_current_valid {
+            Some(Self { current, total })
+        } else {
+            None
+        }
+    }
+
+    /// ID of the shard, 0-indexed.
+    pub const fn current(self) -> u64 {
+        self.current
+    }
+
+    /// Total number of shards, 1-indexed.
+    pub const fn total(self) -> u64 {
+        self.total
+    }
+}
+
+/// Display the shard ID.
+///
+/// Formats as `shard {current}/{total}`.
+impl Display for ShardId {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        f.write_str("shard ")?;
```

Comment:
I think, from looking at the example in `Display`'s documentation, that the type should not be included in the output.
Users will generally include it before calling the `Display` implementation e.g. `println!("shard id: {}", shard.id());`.
```suggestion
```